[RFC] sysroot: Support boot counting for boot entries

igoropaniuk opened the pull request:
```
Add support for boot counting for bootloader entries [1]. The boot counting data is stored in the name of the boot loader entry. A boot loader entry file name may contain a plus (+) followed by a number. This may optionally be followed by a minus (-) followed by a second number. The dot (.) and file name suffix (conf or efi) must immediately follow.

All "pending" entries (in the middle of boot counting) are automatically removed during creation of new boot entries for new deployments.

Testing:
$ ostree admin deploy 91fc19319be9e79d07159303dff125f40f10e5c25614630dcbed23d95e36f907 Copying /etc changes: 2 modified, 3 removed, 4 added bootfs is sufficient for calculated new size: 0 bytes Transaction complete; bootconfig swap: yes; bootversion: boot.0.1, deployment count change: 1

$ ls /boot/loader/entries
ostree-1+3.conf  ostree-2+3.conf

[1] https://uapi-group.org/specifications/specs/boot_loader_specification/#boot-counting
```

cgwalters commented:
I love the idea of integrating here, thanks for starting this! This came up before in at least https://github.com/ostreedev/ostree/issues/3032

Hmm don't we need to consider how we make this configurable? Currently systemd kernel-install parses `/etc/kernel/tries`. (I think we should also support a `/usr/lib` variant of this, but that's mostly an aside). I guess we could in theory just start honoring that file too?

igoropaniuk commented:
@cgwalters thanks for the feedback, my concern regarding parsing `/etc/kernel/tries` is that it's systemd-specific config, which is not described in `https://uapi-group.org/specifications/specs/boot_loader_specification/`

A can add an optional autoconf flag for that

igoropaniuk commented:
@cgwalters added support for `/etc/kernel/tries` as suggested

cgwalters commented on configure.ac:
```diff
@@ -362,6 +362,14 @@ AS_IF([test "$enable_man_html" = yes && test "$enable_man" = no], [
 ])
 AM_CONDITIONAL(ENABLE_MAN_HTML, test "$enable_man_html" = yes)
 
+AC_ARG_ENABLE(boot-count,
```
I think instead of a build-time option this should be a runtime option in the sysroot; something like 
```
[sysroot]
boot-counting = true
```
or so?

igoropaniuk replied:
yes, makes sense, will fix. thanks!

igoropaniuk replied:
fixed

cgwalters commented on src/libostree/ostree-sysroot-deploy.c:
```diff
@@ -1792,14 +1792,78 @@ bootloader_entry_filename (OstreeSysroot *sysroot, guint n_deployments,
+bootloader_get_max_boot_tries (OstreeSysroot *self, OstreeDeployment *deployment,
+                               GCancellable *cancellable, GError **error)
+{
+  g_autofree char *tries_file_path = ostree_deployment_get_boot_tries_relpath (deployment);
```
Hmmm...one super messy thing here is that at this point we may not have actually written out the deployment's `/etc`. We really want a version of this that's in `/usr`.

AFAICS, the uapi standard doesn't specify `/etc/kernel/tries` - that's an invention of `kernel-install`.

To combine with "make this configurable via sysroot" we could have e.g.:

```
[sysroot]
boot_counting=3
```

But OTOH I've been trying to get away from the sysroot configuration for other reasons...so we may need to invent `/usr/lib/ostree/sysroot.conf` or so as a pair with `/usr/lib/ostree/prepare-root.conf`?

igoropaniuk replied:
added support for `boot_counting` property in repo config

cgwalters reviewed (commented):
Thanks so much for working on this!

cgwalters commented:
BTW, which bootloader are you looking at using this with?

igoropaniuk commented:
@cgwalters EDK2 + systemd-boot

igoropaniuk commented:
@cgwalters I've addressed all comments,  thanks!

igoropaniuk commented:
@cgwalters looks like some CI tests are still failing, but all of them are unrelated to this change, for example:
```
2024-10-30T16:06:09.9538403Z ERROR: tests/test-commit-sign.sh - too few tests run (expected 7, got 2)
2024-10-30T16:06:09.9539247Z ERROR: tests/test-commit-sign.sh - exited with status 1
```

cgwalters commented:
Yes CI failures are unrelated, will look at this soon

jlebon commented on man/ostree.repo-config.xml:
```diff
@@ -416,6 +416,14 @@ License along with this library. If not, see <https://www.gnu.org/licenses/>.
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>boot_counting</varname></term>
```
Bikeshed: for consistency with other knobs in this file, `boot-counting`?

Also, should this be e.g. `boot-counting-tries` to make it clearer?

igoropaniuk replied:
Changed to `boot-counting-tries`

jlebon commented on man/ostree.repo-config.xml:
```diff
@@ -416,6 +416,14 @@ License along with this library. If not, see <https://www.gnu.org/licenses/>.
+        <term><varname>boot_counting</varname></term>
+        <listitem><para>Integer value controlling the number of maximum boot attempts. The boot counting data is stored in the name of the
+        boot loader entry. A boot loader entry file name may contain a plus (+) followed by a number. This may optionally be followed by
+        a minus (-) followed by a second number. The dot (.) and file name suffix (conf or efi) must immediately follow.
```
Feel like a link to the spec here would be helpful.

igoropaniuk replied:
Added a link to the sepc

jlebon commented on src/libostree/ostree-repo.c:
```diff
@@ -3297,6 +3297,20 @@ reload_remote_config (OstreeRepo *self, GCancellable *cancellable, GError **erro
+                                                            &boot_counting_str, NULL);
+
+    if (boot_counting_str)
+      /* Ensure boot count value is in [0,5] */
```
Hmm, doesn't seem required by the spec. Is that a systemd-boot requirement?

igoropaniuk replied:
well, I took the initiative here :) . I personally don't see any reasons why we should have support for more then 5 boot attempts.

I can drop this limitation if there are objections.

jlebon commented on src/libostree/ostree-sysroot-deploy.c:
```diff
@@ -1792,14 +1792,72 @@ bootloader_entry_filename (OstreeSysroot *sysroot, guint n_deployments,
+  if (!ostree_sysroot_get_repo (self, &repo, cancellable, error))
+    return FALSE;
+
+  return (repo->boot_counting != 0 ? TRUE : FALSE);
```
```suggestion
  return repo->boot_counting != 0;
```

jlebon commented on src/libostree/ostree-sysroot-deploy.c:
```diff
@@ -2146,8 +2204,22 @@ install_deployment_kernel (OstreeSysroot *sysroot, int new_bootversion,
     return FALSE;
 
+  g_autofree char *bootconf_filename;
+  if (bootloader_is_boot_count_enabled(sysroot, cancellable, error))
+    {
+      guint max_tries = bootloader_get_max_boot_tries (sysroot, cancellable, error);
+      bootconf_filename = g_strdup_printf ("%s+%u.conf", bootconf_name, max_tries);
+
+      if (!bootloader_remove_tmp_entries(bootconf_dfd, bootconf_name, max_tries, cancellable, error))
+        return FALSE;
+    }
+  else
+    {
+      bootconf_filename = g_strdup_printf ("%s.conf", bootconf_name);
+    }
+
   if (!ostree_bootconfig_parser_write_at (ostree_deployment_get_bootconfig (deployment),
-                                          bootconf_dfd, bootconf_name, cancellable, error))
+                                          bootconf_dfd, bootconf_filename, cancellable, error))
```
Hmm, but note that this function is called for _every_ deployment on _every_ update (that results in a BLS write). So this would effectively "unbless" a boot that was previously blessed (or reset to the max boots that did have some failures).

ISTM like the boot count needs to be part of e.g. the `OstreeDeployment` object so that we can leave it untouched when rewriting BLS entries.

igoropaniuk replied:
Technically we can have just a boolean flag (`bootable` field for example) for that, and check if a filename of an old BLS entry (stored in `/boot/loader.%d/entries`) still has a boot counter suffix. And if it doesn't -> set OstreeDeployment->bootable, so then we can create all new BLS entries for this particular deployment without any boot counters.

The problem is that deployment index for OstreeDeployment is not static and there is no deterministic way to match existing boot entry with a specific OstreeDeployment, maybe only parsing BLS entry contents could help

jlebon replied:
> The problem is that deployment index for OstreeDeployment is not static and there is no deterministic way to match existing boot entry with a specific OstreeDeployment, maybe only parsing BLS entry contents could help

There is, but that state is currently not kept around. If you look at e.g. `sysroot_load_from_bootloader_configs()` is where we go through the BLS entries and convert those to an ordered list of `OstreeDeployment`s. So we could parse the counter information from the filename there and keep that state in the deployment objects.

igoropaniuk replied:
Re-implemented as you suggested, please take a look

igoropaniuk commented:
The PR is not stalled, I'll get back to it tomorrow

igoropaniuk commented on src/libostree/ostree-sysroot-deploy.c:
```diff
@@ -4213,14 +4253,35 @@ ostree_sysroot_deployment_set_kargs_in_place (OstreeSysroot *self, OstreeDeploym
+        {
+          guint max_tries = bootloader_get_max_boot_tries (self, cancellable, error);
+
+          if (!ostree_bootconfig_parser_is_parsed (new_bootconfig))
```
Just realized that it makes sense to put that into a separate function to avoid code duplication

igoropaniuk replied:
fixed

igoropaniuk commented:
@jlebon @cgwalters all comments фку addressed, any additional suggestions/objections are welcome

BTW, I've cross-checked CI checks that failed, looks like they are unrelated again

cgwalters commented on src/libostree/ostree-bootconfig-parser.h:
```diff
@@ -67,4 +67,13 @@ void ostree_bootconfig_parser_set_overlay_initrds (OstreeBootconfigParser *self,
 char **ostree_bootconfig_parser_get_overlay_initrds (OstreeBootconfigParser *self);
 
+_OSTREE_PUBLIC
+guint64 ostree_bootconfig_parser_get_tries_left (OstreeBootconfigParser *self);
```
If we want these to actually be public they'll need to be added to the `libostree-devel.sym` file.

cgwalters commented on src/libostree/ostree-bootconfig-parser.c:
```diff
@@ -56,6 +59,94 @@ ostree_bootconfig_parser_clone (OstreeBootconfigParser *self)
+ * ostree_bootconfig_parser_get_tries_left:
+ * @self: Parser
+ *
+ * Returns: Amount of boot tries left
```
Please also add:
`Since: 2025.2` (here and below)

cgwalters commented on src/libostree/ostree-bootconfig-parser.c:
```diff
@@ -56,6 +59,94 @@ ostree_bootconfig_parser_clone (OstreeBootconfigParser *self)
+ * filename (excluding file extension).
+ */
+static void
+ostree_bootconfig_parser_parse_tries (OstreeBootconfigParser *self,  const char *filename)
```
I think this one would be easier to unit test if it was just
```
parse_bootloader_tries (const char *filename, guint32 *out_left, guint32 *out_done, GError **error)
```

(I also added an error so we can error if we fail to parse the numbers which seems like a good idea?)

Then it'd be good to have some unit tests for that. Actually, I will try to do a prep PR to add unit tests for some other code in this file just to start that ball.

cgwalters reviewed (commented):
A few comments, thanks again for this work!

igoropaniuk commented:
@cgwalters Thank you for your review. I have addressed all your recently posted comments. Please let me know if there's anything else I need to do.

cgwalters commented on src/libostree/ostree-sysroot-deploy.c:
```diff
@@ -1797,14 +1798,35 @@ parse_os_release (const char *contents, const char *split)
 }
 
+static guint
+bootloader_get_max_boot_tries (OstreeSysroot *self, GCancellable *cancellable, GError **error)
+{
+  g_autoptr (OstreeRepo) repo = NULL;
+  if (!ostree_sysroot_get_repo (self, &repo, cancellable, error))
+    return 0;
+
+  return repo->boot_counting;
```
This function doesn't adhere to the GError conventions (I know, it's messy) - or at least the caller down below just ignores a return value of `0` and continues (potentially leaving the `error` set).

But I think there's a simple solution; we don't need this wrapper function at all. It's safe to just deref from `self->repo->boot_counting` in callers (if the sysroot is loaded).

igoropaniuk replied:
removed this function

cgwalters commented on src/libostree/ostree-sysroot-deploy.c:
```diff
@@ -1797,14 +1798,35 @@ parse_os_release (const char *contents, const char *split)
+}
+
+static gboolean
+bootloader_is_boot_count_enabled (OstreeSysroot *self, GCancellable *cancellable, GError **error)
```
Ditto for this

igoropaniuk replied:
removed this function

cgwalters commented on src/libostree/ostree-sysroot-deploy.c:
```diff
@@ -1797,14 +1798,35 @@ parse_os_release (const char *contents, const char *split)
-                           OstreeDeployment *deployment)
+                           OstreeDeployment *deployment, GCancellable *cancellable, GError **error)
 {
+  g_autofree char *bootconf_name;
```
Initialize to `NULL` for safety

igoropaniuk replied:
done

cgwalters commented on src/libostree/ostree-sysroot-deploy.c:
```diff
@@ -1797,14 +1798,35 @@ parse_os_release (const char *contents, const char *split)
 static char *
 bootloader_entry_filename (OstreeSysroot *sysroot, guint n_deployments,
-                           OstreeDeployment *deployment)
+                           OstreeDeployment *deployment, GCancellable *cancellable, GError **error)
```
After we drop the fallibility (use of GError) from those helpers we don't need this change either.

igoropaniuk replied:
sure, dropped these params

cgwalters commented on src/libostree/ostree-bootconfig-parser.c:
```diff
@@ -56,6 +59,96 @@ ostree_bootconfig_parser_clone (OstreeBootconfigParser *self)
+  *out_left = 0;
+  *out_done = 0;
+
+  for (;;)
+    {
+      char *plus = strchr (counter ?: filename, '+');
+      if (plus)
+        {
+          /* Look for the last "+" symbol in the filename */
+          counter = plus + 1;
+          continue;
+        }
+      if (counter)
+        break;
+
+      /* No boot counter found */
+      return;
+    }
```
I had to stare at this a bit...I think it's correct as is, but wouldn't it be a lot clearer to use `strrchr ()` here (note the extra `r` to search from the end)?

i.e.
```
const char *counter = strrchr (filename, '+');
if (!counter)
  return;
counter += 1;
```

igoropaniuk replied:
fixed, thanks!

cgwalters commented on src/libostree/ostree-bootconfig-parser.c:
```diff
@@ -56,6 +59,96 @@ ostree_bootconfig_parser_clone (OstreeBootconfigParser *self)
+static void
+parse_bootloader_tries (const char *filename, guint64 *out_left, guint64 *out_done, GError **error)
+{
+  gchar *counter = NULL;
+  gchar *old_counter = NULL;
```
Since these are pointing into `filename` they should also be `const`

cgwalters commented on src/libostree/ostree-bootconfig-parser.c:
```diff
@@ -56,6 +59,96 @@ ostree_bootconfig_parser_clone (OstreeBootconfigParser *self)
+
+  guint64 tries_left, tries_done = 0;
+
+  old_counter = counter;
+  tries_left = g_ascii_strtoull (old_counter, &counter, 10);
```
I think this is also OK but it would seem clearer to me to leave the value of `counter` pointing where it is and then assign to a new `tries` variable? Like:

```
const char *tries = NULL;
tries_left = g_ascii_strtoull (counter, &tries, 10);
...
```
or so?

igoropaniuk replied:
I quickly looked into that and created a prototype, but it seems to only make the code more complex IMO

cgwalters reviewed (changes requested):
Thanks so much for working on this! I added some comments but I don't mind helping out myself by pushing updates here if you prefer, let me know!

One thing that I think would be helpful is some new unit tests, but I am also happy to do that.

igoropaniuk commented:
@cgwalters Apologies for the delay. I've addressed all the comments except the last one concerning the "usage" variable.

Could we get this PR to land first (as it seems closer to being merged in terms of its maturity), and then I'll focus on addressing all issues in the other one (https://github.com/ostreedev/ostree/pull/3310) afterwards?

cgwalters commented:
/gemini review

cgwalters commented:
OK I

- Addressed the gemini review comments
- Added unit tests for the tries parsing
- Added a unit-admin test that verifies we write the bootloader entry files w/boot counting initalized

cgwalters commented:
/gemini review

cgwalters commented on src/libostree/libostree-devel.sym:
```diff
@@ -1,16 +1,16 @@
   License as published by the Free Software Foundation; either
   version 2 of the License, or (at your option) any later version.
- 
+
```
these are spurious changes, we should remove them

cgwalters commented on src/libostree/ostree-bootconfig-parser.h:
```diff
@@ -67,4 +67,13 @@ void ostree_bootconfig_parser_set_overlay_initrds (OstreeBootconfigParser *self,
+_OSTREE_PUBLIC
+guint64 ostree_bootconfig_parser_get_tries_left (OstreeBootconfigParser *self);
+
+_OSTREE_PUBLIC
```
Why are these public APIs? I'm not saying it's wrong, but it's not immediately obvious to me why they should be.

Well, I guess actually we probably should be rendering this in e.g. `ostree admin status` at least.

But right now only `_tries_left` is listed in the symbols file, so if we're making these public we should ensure that `_tries_done` is too.

cgwalters commented on src/libostree/ostree-bootconfig-parser.c:
```diff
@@ -56,6 +59,87 @@ ostree_bootconfig_parser_clone (OstreeBootconfigParser *self)
+ * Returns: TRUE if the boot config was parsed from existing boot config file
+ */
+gboolean
+ostree_bootconfig_parser_is_parsed (OstreeBootconfigParser *self)
```
Hmm how about storing the original filename instead? i.e. `ostree_bootconfig_parser_get_filename`. It seems more generally useful and easy to describe than what "is_parsed" means.

cgwalters reviewed (changes requested):
Forgot to also mention I went ahead and rebased this on git main.

I have just a few other minor comments but I think we can get this in by tomorrow - what do you think?

igoropaniuk commented:
@cgwalters, it seems we are simultaneously working to address comments.
Would you like me to finish addressing all the issues, or would you like to handle that on your own?

cgwalters commented:
> @cgwalters, it seems we are simultaneously working to address comments :)

Ahhh sorry! Based on your location I thought you'd be sleeping now :smile: 

OK I did just one more force push but I will hand over the PR to you! What do you think of the changes?

igoropaniuk commented:
@cgwalters 

> Ahhh sorry! Based on your location I thought you'd be sleeping now 😄
> OK I did just one more force push but I will hand over the PR to you! What do you think of the changes?

Great, thanks for covering this functionality with unit tests and addressing the reported issues by Gemini!
Please let me know if you have any additional comments; I'll review them tomorrow morning and try to address them right away + address build issues reported by CI as well

cgwalters commented:
Offhand I don't have any other issues!

It looks like I broke CI by adding `ostree_bootconfig_parser_get_tries_done` to public API without also listing it in `apidoc/` - that should be an easy fix.

igoropaniuk commented:
> It looks like I broke CI by adding ostree_bootconfig_parser_get_tries_done to public API without also listing it in apidoc/ - that should be an easy fix.

Fixed, thanks for the hint!

cgwalters commented:
/override ci/prow/fcos-e2e
I'll probably drop this one as it's always duplicated coreos ci jenkins
